pkg/certificate/signer.go

```diff
@@ -0,0 +1,528 @@
+
+	var err error
+
+	if config.RestMapper == nil {
+		if s.restMapper, err = util.BuildRestMapper(config.RestConfig); err != nil {
+			return nil, errors.Wrap(err, "error building the REST mapper")
+		}
+	}
+
+	if config.DynClient == nil {
+		if s.dynClient, err = resource.NewDynamicClient(config.RestConfig); err != nil {
+			return nil, errors.Wrap(err, "error creating dynamic client")
+		}
+	}
+
+	return s, nil
+}
+
+func (s *signerImpl) Start(namespace string) error {
+	if _, exists := s.syncerMap.Load(namespace); exists {
+		return nil
+	}
+
+	secretSyncer, err := syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:            "Cert Signer",
+		SourceClient:    s.dynClient,
+		SourceNamespace: namespace,
+		RestMapper:      s.restMapper,
+		ResourceType:    &corev1.Secret{},
+		ShouldProcess: func(obj *unstructured.Unstructured, op syncer.Operation) bool {
+			return op != syncer.Delete && obj.GetLabels()[SigningRequestLabelKey] != "" && obj.GetAnnotations()[RequestSignedLabelKey] == ""
+		},
+		Transform: func(from runtime.Object, _ int, _ syncer.Operation) (runtime.Object, bool) {
+			secret := from.(*corev1.Secret)
+
+			err := s.signSecret(secret)
+			if err != nil {
+				logger.Errorf(err, "error signing Secret %q", secret.Name)
+			}
+
+			return secret, err != nil
+		},
+		Federator: federate.NewUpdateFederator(s.dynClient, s.restMapper, namespace,
+			func(oldObj *unstructured.Unstructured, newObj *unstructured.Unstructured) *unstructured.Unstructured {
+				existingSecret := resource.MustFromUnstructured(oldObj, &corev1.Secret{})
+				updatedSecret := resource.MustFromUnstructured(newObj, &corev1.Secret{})
+
+				if existingSecret.Annotations == nil {
+					existingSecret.Annotations = map[string]string{}
+				}
+
+				existingSecret.Annotations[RequestSignedLabelKey] = "true"
+				existingSecret.Data[TLSDataKey] = updatedSecret.Data[TLSDataKey]
+				existingSecret.Data[CADataKey] = updatedSecret.Data[CADataKey]
+
+				return resource.MustToUnstructured(existingSecret)
+			}),
+	})
+	if err != nil {
+		return errors.Wrap(err, "error creating resource syncer")
+	}
+
+	stopCh := make(chan struct{})
+
+	if err := secretSyncer.Start(stopCh); err != nil {
+		return errors.Wrap(err, "error starting resource syncer")
+	}
+
+	s.syncerMap.Store(namespace, stopCh)
+
+	// Start periodic CA check if not already running
+	s.startPeriodicCACheck(namespace)
+
+	return nil
+}
+
+func (s *signerImpl) Stop(namespace string) {
+	if v, exists := s.syncerMap.LoadAndDelete(namespace); exists {
+		close(v.(chan struct{}))
+	}
+
+	// Stop periodic CA check when stopping the signer
+	s.stopPeriodicCACheck()
+}
+
+func (s *signerImpl) signSecret(secret *corev1.Secret) error {
+	csrPEM := secret.Data[CSRDataKey]
+	if len(csrPEM) == 0 {
+		return fmt.Errorf("CSR data is empty")
+	}
+
+	// Get CA secret using dynamic client
+	caSecretClient := s.dynClient.Resource(corev1.SchemeGroupVersion.WithResource("secrets")).Namespace(s.namespace)
```

Comment:
Since we have signers per namespace, we want to use the `Namespace` from the `secret`.

```suggestion
	caSecretClient := s.dynClient.Resource(corev1.SchemeGroupVersion.WithResource("secrets")).Namespace(secret.Namespace)
```